error_writer.go

```diff
@@ -87,9 +89,15 @@ func NewErrorWriter(opts ...HandlerOption) *ErrorWriter {
 func (w *ErrorWriter) classifyRequest(request *http.Request) protocolType {
 	ctype := canonicalizeContentType(getHeaderCanonical(request.Header, headerContentType))
 	if _, ok := w.unaryConnectContentTypes[ctype]; ok {
+		if err := connectCheckProtocolVersion(request, w.requireConnectProtocolHeader); err != nil {
+			return unknownProtocol
+		}
 		return connectUnaryProtocol
 	}
 	if _, ok := w.streamingConnectContentTypes[ctype]; ok {
+		if err := connectCheckProtocolVersion(request, w.requireConnectProtocolHeader); err != nil {
+			return unknownProtocol
+		}
```

Comment:
This doesn't feel right. Even if the server config requires a connect protocol version header, the content-type alone makes it clear that this _is_ the Connect streaming protocol. Only unary RPCs have ambiguity, due to use of "application/proto" and "application/json" content-types (which are not necessarily Connect calls).

@akshayjshah, what do you think?

Reply:
Ambiguity could arise for a v2 connect spec which we'd want older clients to protect against by failing on the version check. `unknown` is probably not the right error for this case.

Reply:
It still wouldn't be ambiguous. If/when we introduce a v2, the version header will _have_ to become required in the spec, which means a missing header _implies_ v1 of the protocol. So we still know the format for the error, regardless of the presence of the header.

The only ambiguity is between Connect unary and non-Connect responses that use the same content types (most notably "application/json").

Reply:
@jhump, I think you've nailed it: we should use _different_ header-checking logic here. If the version header is present and indicates v1, use the current Connect streaming encoding. If the version header is missing, also use the current Connect streaming encoding. If the version header is present and _not_ set to v1, we've gotten a request using a version from the future: treat it as an unknown protocol. Regardless, we don't need to worry about whether the RPC server requires the header or not.

As stated in #699 and the , the header is really only useful to distinguish unary Connect requests from non-RPC requests which might use the same Content-Type.

Reply:
Set `required = false` for streaming ensuring we still check the version header is correct.